openpifpaf/network/basenetworks.py

```diff
@@ -498,102 +498,160 @@ def configure(cls, args: argparse.Namespace):
         cls.pretrained = args.squeezenet_pretrained
 
 
+class FPN(torch.nn.Module):
```

Comment:
OpenPifPaf already depends on `torchvision` which does have a FeaturePyramidNetwork implementation. Could this be used here? If not, add a small comment regarding this in your docstring. You had it in the PR comment, but please also add the link to the paper here (just to clarify that you are implementing the original FPN).

Reply:
[torchvision's FPN](https://github.com/pytorch/vision/blob/master/torchvision/ops/feature_pyramid_network.py) refines and outputs all feature maps, which means applying a 3x3 conv on all of the feature maps, even if they don't get used.

Here, this implementation only refines and outputs the desired level, which means that if, for example, I want the feature map of stride 16 from the 3rd FPN level, I won't need to also compute the feature maps of stride 4, 8 and 32.

While it should be possible to avoid refining higher res feature maps with torchvision's FPN by adapting the codebase here, I don't see a way to avoid refining the lower res feature maps. The changes to avoid refining higher res feature maps would also complexify the `__init__` and `forward` of the `SwinTransformer` class, but I think that's minor compared to the impact on compute. 

I updated the docstring and added a link to the paper, but do let me know if you think it's worth switching to torchvision's FeaturePyramidNetwork @svenkreiss .

Reply:
That all makes sense to me. Thanks for clarifying! Do you mind putting the first two paragraphs (or something similar) into the docstring? I am pretty sure someone (and it might be me) will have the same question again in six months and the PR discussions here are usually lost.